### ISSUE TITLE
[New Rule] Static methods SHOULD NOT be used

### ISSUE THREAD

**Issue by lenaorobei** [magento/magento-coding-standard#24]
### Rule
The use of `static` methods is discouraged.

### Reason
Source: [Magento Technical Guidelines](https://devdocs.magento.com/guides/v2.3/coding-standards/technical-guidelines.html).

[Plugins](https://devdocs.magento.com/guides/v2.0/extension-dev-guide/plugins.html#plugin-limit) cannot be used with static methods, because static methods cannot be intercepted. 
As a result, It hurts extensibility.

### Implementation
- Subscribe to the `T_STATIC ` token.
- If `T_FUNCTION ` is found right to it raise a warning.

**Comment by larsroettig** [comment]
Hi i will work on this at next weekend

**Comment by joni-jones** [comment]
This rule should not include tests which use PHPUnit because all `assert...`, `once()`, `atLeast()`, etc. are static methods.

**Comment by paliarush** [comment]
@joni-jones I believe this is not relevant because the proposed sniff is looking for declaration of the static tests, not usage. Additionally, in the recent versions of PHPUnit all this methods are not invoked statically.

**Comment by joni-jones** [comment]
> Additionally, in the recent versions of PHPUnit all this methods are not invoked statically.

This issue contains the explanation https://github.com/sebastianbergmann/phpunit/issues/1914 and you can see Sebastian's answer and the community reaction.

**Comment by lenaorobei** [comment]
Implemented in magento/magento-coding-standard#24

**Comment by hopeseekr** [comment]
The PHPUnit maintainer is simply wrong on this issue. I have thoroughly analyzed PHPUnit's code, and unless we are going to declare willfully disregarding the static method calls are OK, then such an important code quality tool as PHPUnit should have code written for it using `self::assertEquals()` vs. the contextually ambiguous `$this->assertEquals()`.

Or, PHPUnit needs to have a third class that does something like

```php
// Expose static assert methods for people who need them.
class Assert
{
    public static function assertEquals($expected, $actual, $memo='')
    {
        return (new AssertLogic())->assertEquals($expected, $actual, $memo);
    }
}
```

Then extend `TestCase` against `AssertLogic`. Then literally everyone can be happy, at probably minimal overhead. (if the news are too much, then `private static $asserter`.
